src/blk/aio/aio.cc

```diff
@@ -25,33 +25,43 @@ int aio_queue_t::submit_batch(aio_iter begin, aio_iter end,
   int r;
 
   aio_iter cur = begin;
-  struct aio_t *piocb[aios_size];
-  int left = 0;
-  while (cur != end) {
-    cur->priv = priv;
-    *(piocb+left) = &(*cur);
-    ++left;
-    ++cur;
-  }
-  ceph_assert(aios_size >= left);
+#if defined(HAVE_LIBAIO)
+  struct aio_t *piocb[max_iodepth];
+#endif
   int done = 0;
-  while (left > 0) {
+  int pushed = 0; //used for LIBAIO only
+  int pulled = 0;
+  while (cur != end || pushed < pulled) {
 #if defined(HAVE_LIBAIO)
-    r = io_submit(ctx, std::min(left, max_iodepth), (struct iocb**)(piocb + done));
+    while (cur != end && pulled < max_iodepth) {
+      cur->priv = priv;
+      piocb[pulled] = &(*cur);
+      ++pulled;
+      ++cur;
+    }
+    int toSubmit = pulled - pushed;
+    r = io_submit(ctx, toSubmit, (struct iocb**)(piocb + pushed));
+    if (r >= 0 && r < toSubmit) {
+      pushed += (r > 0 ? r : 0);
```

Comment:
I don't see a point in the `(r > 0)` comparisons. As we are in the if's body, `r` must be already `0` or positive. However, the ternary operator for `r == 0` yields... `0` IIUC.

Reply:
right, my bad. thanks!